core/src/main/java/org/infinispan/interceptors/TxInterceptor.java

```diff
@@ -235,18 +249,73 @@ public Object visitSizeCommand(InvocationContext ctx, SizeCommand command) throw
    }
 
    @Override
-   public Object visitKeySetCommand(InvocationContext ctx, KeySetCommand command) throws Throwable {
-      return enlistReadAndInvokeNext(ctx, command);
-   }
-
-   @Override
-   public Object visitValuesCommand(InvocationContext ctx, ValuesCommand command) throws Throwable {
-      return enlistReadAndInvokeNext(ctx, command);
+   public CacheSet<K> visitKeySetCommand(InvocationContext ctx, KeySetCommand command) throws Throwable {
+      CacheSet<K> set = (CacheSet<K>) enlistReadAndInvokeNext(ctx, command);
+      if (ctx.isInTxScope()) {
```

Comment:
Unfortunately this usage is desired.  Size for example is a big one that comes up in the middle of a transaction I have found.